alibi_detect/od/knn.py

```diff
@@ -0,0 +1,130 @@
+        normalizer: Optional[Union[TransformProtocol, FittedTransformProtocol]] = None,
+        aggregator: Optional[TransformProtocol] = None,
+        backend: Literal['pytorch'] = 'pytorch'
+    ) -> None:
+        """
+        k-Nearest Neighbours (kNN) outlier detector.
+
+        Parameters
+        ----------
+        k
+            Number of nearest neighbours to use for outlier detection. If an array is passed, an aggregator is required
+            to aggregate the scores.
+        kernel
+            Kernel function to use for outlier detection. If None, `torch.cdist` is used.
+        normalizer
+            Normalizer to use for outlier detection. If None, no normalisation is applied.
+        aggregator
+            Aggregator to use for outlier detection. If None, no aggregation is applied. If an array is passed for `k`,
+            then an aggregator is required.
+        backend
+            Backend used for outlier detection. Defaults to `'pytorch'`. Options are `'pytorch'`.
+
+        Raises
+        ------
+        ValueError
+            If `k` is an array and `aggregator` is None.
+        NotImplementedError
+            If choice of `backend` is not implemented.
+        """
+        super().__init__()
+
+        backend_str: str = backend.lower()
+        BackendValidator(
+            backend_options={'pytorch': ['pytorch']},
+            construct_name=self.__class__.__name__
+        ).verify_backend(backend_str)
+
+        if isinstance(k, (list, np.ndarray)) and aggregator is None:
+            raise ValueError((f'k={k} is type {type(k)} but aggregator is {aggregator}, you must '
+                              'specify at least an aggregator if you want to use the knn detector '
+                              'ensemble like this.'))
+
+        backend_cls, accumulator_cls = backends[backend]
+        accumulator = None
+        if normalizer is not None or aggregator is not None:
+            accumulator = accumulator_cls(
+                normalizer=normalizer,
+                aggregator=aggregator
+            )
+        self.backend = backend_cls(k, kernel=kernel, accumulator=accumulator)
+
+    def fit(self, x_ref: Union[np.ndarray, List]) -> None:
+        """Fit the kNN detector on reference data.
+
+        Parameters
+        ----------
+        x_ref
+            Reference data used to fit the kNN detector.
+        """
+        self.backend.fit(self.backend._to_tensor(x_ref))
+
+    def score(self, X: Union[np.ndarray, List]) -> np.ndarray:
+        """Score X instances using the kNN.
+
+        Parameters
+        ----------
+        X
+            Data to score. The shape of `X` should be `(n_instances, n_features)`.
+
+        Returns
+        -------
+        Anomaly scores. The shape of the scores is `(n_instances,)`. The higher the score, the more anomalous the \
+        instance.
+        """
+        score = self.backend.score(self.backend._to_tensor(X))
+        return score.numpy()
```

Comment:
I've implemented this as a method `_to_numpy` on the backend. It basically just does: `x.cpu().detach().numpy()` and returns the `np.ndarray`. This can be implemented separately for different backends. Similarly, the `TorchOutlierDetectorOutput` has a `_to_numpy` method that transforms the prediction outputs to numpy arrays. I'm thinking this is a little strange though, it feels like I'm doing the same thing in different places using different objects and there is probably a more consistent approach. I'll look into this further.

Reply:
I've updated the `_to_numpy` method on the base torch outlier detector class so that it converts both `torch tensors to `numpy` array and converts any `torch` tensors found in `TorchOutlierDetectorOutput` to `numpy` arrays. This means that all of these conversions happen in the same place on the torch backend.

Not super confident in this approach. LMK what you think?